util/overlay/overlay_linux.go

```diff
@@ -126,7 +127,14 @@ func WriteUpperdir(ctx context.Context, w io.Writer, upperdir string, lower []mo
 	}
 	return mount.WithTempMount(ctx, lower, func(lowerRoot string) error {
 		return mount.WithTempMount(ctx, upperView, func(upperViewRoot string) error {
-			cw := archive.NewChangeWriter(&cancellableWriter{ctx, w}, upperViewRoot)
+			var opts []archive.ChangeWriterOpt
+			if sourceDateEpoch != nil {
+				opts = append(opts,
+					archive.WithModTimeUpperBound(*sourceDateEpoch),
+					archive.WithWhiteoutTime(*sourceDateEpoch),
```

Comment:
Why make the whiteout time configurable at all? These files never end up in a container, so the timestamp value is never visible. Lets just always make it zero.

If you agree, then this part can be separated out as no blob tracking etc is needed for that.

Reply:
Unfortunately not possible with the current version of containerd (and I don't want this PR to be blocked until containerd v2.0), as https://pkg.go.dev/github.com/containerd/containerd@v1.7.2/diff#Opt lacks a knob to set whiteout time without setting `SOURCE_DATE_EPOCH`.

Reply:
Not sure I understand. If containerd doesn't support it, why can't we just call `archive.WithWhiteoutTime(zeroTime)` in here?

Reply:
The problem is that we can't specify `WithWhiteoutTime` for `Differ.Compare()`, although we can pass `WithSourceDateEpoch` to it.
https://github.com/moby/buildkit/blob/2ff0d2a2f53663aae917980fa27eada7950ff69c/cache/blobs.go#L191-L214
https://pkg.go.dev/github.com/containerd/containerd@v1.7.2/diff#Comparer
https://pkg.go.dev/github.com/containerd/containerd@v1.7.2/diff#WithSourceDateEpoch

Reply:
@tonistiigi 🙏

Reply:
So atm we could only fix it for the overlay differ but not for containerd client based differ for other snapshotters?

Reply:
I don't think we have an overlay-specific implementation of https://pkg.go.dev/github.com/containerd/containerd@v1.7.2/diff#Comparer , so I'm not sure.

Either way, I'd prefer to keep the behavior as much consistent as possible across the snapshotters.

Reply:
BuildKit diffing goes always though `GetRemotes->computeBlobChain->tryComputeOverlayBlob`. Comparer containerd API is never reached in the happy path.

Reply:
👍 , but what should we do this for now?
Do we have to give up reproducibility for non-overlay snapshotters?

I'd still prefer to maximize reproducibility across snapshotters though (with non-zero whiteout timestamps)

Reply:
There are two different solutions in this PR.
- Rewriting timestamps of user's files in the differ component
- Making whiteout timestamps reproducible (that can not currently be controlled at all).

The first one is the complicated one that has many considerations to the UX, cache system, compression variants etc. It needs more validation/review, and I'm not sure if we can get it into the upcoming release as is.

The whiteout problem is much less controversial. Everyone agrees that whiteout times should not be random. But I would make a case that it should just be always zero and not defined by epoch value. To do that for all containerd snapshotters, it seems more work is needed in containerd side. So options would be to do it only for the overlay happy path for now and follow up with support for other snapshotters when containerd fixes have landed. Or set it to zero for overlay, and for the other snapshotters let it be controlled by epoch while defining clearly that this is temporarily the best effort that will change in the future when containerd side fixes have landed. I don't think it makes sense to set it to epoch for everyone atm if we already know that this is not the solution we want long term.

Reply:
Opened https://github.com/moby/buildkit/pull/3981 for the second one, but still not really beneficial due to https://github.com/moby/buildkit/pull/3980

Reply:
>The first one is the complicated one that has many considerations to the UX, cache system, compression variants etc. It needs more validation/review, and I'm not sure if we can get it into the upcoming release as is.

Would it be mergable if we add a new build arg like `BUILDKIT_SOURCE_DATE_EPOCH_MODE=(conservative|aggressive)` ? (in the labs channel?)

Reply:
> Would it be mergable if we add a new build arg like BUILDKIT_SOURCE_DATE_EPOCH_MODE=(conservative|aggressive) ? (in the labs channel?)

This also makes changes to the blobs reference counting and core function signatures in cache pkg, so isn't ideal for an experimental/unsafe feature because it is not something that lives on its own and can be easily removed.

Reply:
Shall we just make this an exporter option?

Reply:
↑ @tonistiigi WDYT?
The idea is to introduce an image exporter option like `force-epoch`.
The implementation will be similar to `force-compression`.

Reply:
- https://github.com/moby/buildkit/pull/4057